encrypt/encrypt.go

```diff
@@ -27,14 +36,271 @@ var ArgHelp = `
+	for _, file := range files {
+		// check that the input file exists
+		fileInfo, err := os.Stat(file.unencrypted)
+		if err != nil {
+			return err
+		}
+		// check that the input file isn't a directory
+		if fileInfo.IsDir() {
+			return fmt.Errorf("%s is a directory", file.unencrypted)
+		}
+		// check that the output file doesn't exist
+		_, err = os.Stat(file.encrypted)
+		if err == nil {
+			return fmt.Errorf("outfile %s already exists", file.encrypted)
+		}
+	}
+	return nil
+}
+
+// Calculates md5 and sha256 hashes for the unencrypted and encrypted file
+func calculateHashes(fileSet encryptionFileSet) (*hashSet, error) {
+
+	hashes := hashSet{"", "", "", ""}
+
+	// open infile
+	inFile, err := os.Open(fileSet.unencrypted)
+	if err != nil {
+		return nil, err
+	}
+	defer inFile.Close()
+
+	// unencrypted md5
+	md5Hash := md5.New()
+	_, err = io.Copy(md5Hash, inFile)
+	if err != nil {
+		return nil, err
+	}
+	hashes.unencryptedMd5 = hex.EncodeToString(md5Hash.Sum(nil))
+
+	// reset file
+	_, err = inFile.Seek(0, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	// unencrypted sha256
+	shaHash := sha256.New()
+	_, err = io.Copy(shaHash, inFile)
+	if err != nil {
+		return nil, err
+	}
+	hashes.unencryptedSha256 = hex.EncodeToString(shaHash.Sum(nil))
+
+	// open outfile
+	outFile, err := os.Open(fileSet.encrypted)
+	if err != nil {
+		return nil, err
+	}
+	defer outFile.Close()
+
+	// encrypted md5
+	md5Hash.Reset()
+	_, err = io.Copy(md5Hash, outFile)
+	if err != nil {
+		return nil, err
+	}
+	hashes.encryptedMd5 = hex.EncodeToString(md5Hash.Sum(nil))
+
+	// reset file
+	_, err = outFile.Seek(0, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	// encrypted sha256
+	shaHash.Reset()
+	_, err = io.Copy(shaHash, outFile)
+	if err != nil {
+		return nil, err
+	}
+	hashes.encryptedSha256 = hex.EncodeToString(shaHash.Sum(nil))
+
+	return &hashes, nil
+}
+
+// Reads a public key file from a file using the crypt4gh keys module
+func readPublicKey(filename string) (*[32]byte, error) {
+	log.Info("Reading Public key file")
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	publicKey, err := keys.ReadPublicKey(file)
+	if err != nil {
+		return nil, err
+	}
+	return &publicKey, nil
+}
+
+// Generates a crypt4gh key pair, returning only the private key, as the
+// public key isn't needed for encryption.
+func generatePrivateKey() (*[32]byte, error) {
+	log.Info("Generating encryption key")
+
+	_, privateKey, err := keys.GenerateKeyPair()
+	if err != nil {
+		return nil, err
+	}
+	return &privateKey, nil
+}
+
+// Encrypts the data from `filename` into `outFilename` for the given `pubKey`,
+// using the given `privateKey`.
+func encrypt(filename, outFilename string, pubKey, privateKey [32]byte) error {
+	// check if outfile exists
+	if _, err := os.Stat(outFilename); err == nil {
+		return fmt.Errorf("outfile %s already exists", outFilename)
+	}
+
+	// read infile
+	inFile, err := os.Open(filename)
+	if err != nil {
+		return err
```

Comment:
this kinds of errors are really annoying since you as a user don't get any information on `which` file it was that could not be opened.

Reply:
Ah! :) I honestly thought the default error would include that. I'll make the error messages better!

Reply:
Sometimes it does, but when it doesn't it is really annoying.

Reply:
Yet another reason to write the tests at the same time as the funcitons. ;)